File: src/com/keguoyu/minijvm/runtime/data/JvmThread.java
package com.keguoyu.minijvm.runtime.data;

import com.keguoyu.minijvm.runtime.JvmStack;

/**
 * JvmThread表示一个线程
 * pc表示要执行的下一条指令 jvmStack表示虚拟机栈
 */
public class JvmThread {
    //执行完指令之后会更新一次
    private int pc;
    private final JvmStack jvmStack;

    public JvmThread() {
        this.pc = 0;
        this.jvmStack = new JvmStack(1024);
    }

    public void pushFrame(StackFrame stackFrame) {
        stackFrame.attachThread(this);
        jvmStack.push(stackFrame);
    }

    public StackFrame popFrame() {
        return jvmStack.pop();
    }

    public StackFrame current() {
        return jvmStack.currentFrame();
    }

    public void setPc(int pc) {
        this.pc = pc;
    }

    public int getPc() {
        return pc;
    }
}

File: vm/JvmThread.py
#!/usr/bin/env python
# encoding: utf-8

from vm.runtime.Method import Method


class JvmThread:
    def __init__(self):
        from vm.JvmStack import JvmStack
        self.pc = 0
        self.stack = JvmStack(1024)

    def push_frame(self, frame):
        self.stack.push(frame)

    def pop_frame(self):
        return self.stack.pop()

    @property
    def current_frame(self):
        return self.stack.top()

    def new_frame(self, method: Method):
        from vm.StackFrame import StackFrame
        return StackFrame(self, method)

    @property
    def top_frame(self):
        return self.stack.top()

    def is_stack_empty(self):
        return self.stack.is_empty()

    def clear_stack(self):
        self.stack.clear()

    def get_frames(self):
        return self.stack.get_frames()
